lib/TH/generic/THStorageCopy.c

```diff
@@ -15,22 +15,31 @@ void THStorage_(copy)(THStorage *storage, THStorage *src)
   THStorage_(rawCopy)(storage, src->data);
 }
 
-
 #define IMPLEMENT_THStorage_COPY(TYPENAMESRC) \
 void THStorage_(copy##TYPENAMESRC)(THStorage *storage, TH##TYPENAMESRC##Storage *src) \
 { \
-  ptrdiff_t i; \
   THArgCheck(storage->size == src->size, 2, "size mismatch"); \
-  for(i = 0; i < storage->size; i++) \
-    storage->data[i] = (real)src->data[i]; \
+  if(THTypeIdx_(Real) == THTypeIdx_(TYPENAMESRC)) {                   \
+    memcpy(storage->data, src->data, sizeof(real)*storage->size);   /* cast just removes compiler warning */ \
+  } else {                                                              \
+    ptrdiff_t i;                                                        \
+    for(i = 0; i < storage->size; i++)                                  \
+      storage->data[i] = (real)src->data[i];                            \
+    }                                                                   \
 }
 
+#ifndef TH_REAL_IS_HALF
 IMPLEMENT_THStorage_COPY(Byte)
 IMPLEMENT_THStorage_COPY(Char)
 IMPLEMENT_THStorage_COPY(Short)
 IMPLEMENT_THStorage_COPY(Int)
 IMPLEMENT_THStorage_COPY(Long)
 IMPLEMENT_THStorage_COPY(Float)
 IMPLEMENT_THStorage_COPY(Double)
+#else
+/* only allow pass-through for Half */
+IMPLEMENT_THStorage_COPY(Half)
```

Comment:
Right, it was not there. Now it is: 
th> x=torch.HalfStorage(3)
                                                                      [0.0001s]
th> x
 0
 0
 0
[torch.HalfStorage of size 3]

                                                                      [0.0003s]